core/decorators.py

```diff
@@ -0,0 +1,177 @@
+from viur.core import errors, current
+from typing import Callable, Dict, Union, List
+
+import functools
+import logging
+import types
+import warnings
+
+
+def ensure_viur_flags(func: Callable) -> None:
+    try:
+        func.viur_flags
+    except AttributeError:
+        func.viur_flags = {}
+
+
+def access(*access: str | list[str]) -> Callable:
+    """Decorator, which performs the authentication and authorization check.
+
+    To expose a method only to logged in users with the access
+    "root" or ("admin" and "file-edit") or "maintainer"
+    use this decorator like this:
+
+    .. code-block:: python
+        from viur.core.decorators import access
+        @access("root", ["admin", "file-edit"], ["maintainer"])
+        def yourMethod(self):
+            return "You're allowed!"
+    """
+
+    def outer_wrapper(func: Callable):
+        ensure_viur_flags(func)
+
+        func.viur_flags["access"] = access
+
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            user = current.user.get()
+            if not user:
+                raise errors.Unauthorized()
+
+            for acc in access:
+                if isinstance(acc, str):
+                    acc = [acc]
+                assert isinstance(acc, (tuple, list, set))
+
+                missing_access = set(acc).difference(user["access"])
+                if not missing_access:
+                    return func(*args, **kwargs)
+
+            logging.error("%s requires access %s", func.__name__, " OR ".join(map(repr, access)))
+            raise errors.Forbidden()
+
+        return wrapper
+
+    assert access, "No rules set"
+    return outer_wrapper
+
+
+def force_ssl(func: Callable) -> Callable:
+    """
+        Decorator, which forces usage of an encrypted Channel for a given resource.
+        Has no effect on development-servers.
+    """
+    ensure_viur_flags(func)
+    func.viur_flags["ssl"] = True
+    return func
+
+
+def require_skey(func=None, *, allow_empty: bool = False, forward_argument: str = "", **extra_kwargs: dict) -> Callable:
+    """
+    Decorator, which marks the function requires a skey.
+    """
+
+    def decorator(func: Callable) -> Callable:
+        ensure_viur_flags(func)
+
+        func.viur_flags["skey"] = {
+            "allow_empty": allow_empty,
+            "forward_argument": forward_argument,
+            "kwargs": extra_kwargs,
+        }
+
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            if not allow_empty and 'skey' not in kwargs:
+                raise errors.PreconditionFailed()
+            return func(*args, **kwargs)
```

Comment:
Where is the validation for skey? 

You removed
```py
if not securitykey.validate(skey):
    raise errors.PreconditionFailed()
```
in the prototypes, but didn't added it in this decorator.

I suggest to keep the other existence check and add another validation check like
```
if not securitykey.validate(skey):
    raise errors.PreconditionFailed("Invalid skey submitted")
```
This makes it easier for the client to recognize if the skey was not sent at all or if it is wrong (e.g. expired).

Reply:
> Where is the validation for skey?

This is now done in the request routing, as shown yesterday.